### PR TITLE
Homework 5 

### PR DESCRIPTION
**Pull request by asish918**
# Homework 5 Asish

### Topics Covered:
- Multi Threading
- Executor Service
- Synchronization
- Locking and Semaphore uses

### Drift Video Explanation
[Drift](https://video.drift.com/v/abYbUWWfUICx0mkXPwhrnN5nSnwEUHkb4XaUCyN1sF94/)

### Output Screenshots - 

Q1
Screenshot 2024-01-13 at 11 42 36

Q2
Screenshot 2024-01-13 at 11 42 45

Q3
Screenshot 2024-01-13 at 11 42 56

### The bugs mentioned in Sonarqube are weird. It says I should re-throw the Interrupted Exception even though I have thrown it, catched it and nicely logged it. Idk why is that happening 🙃

### CODE REVIEW

**Comment by asish918** [comment]
| Metric | Value |
| ------ | ----- |
| alert_status | <b>:white_check_mark: OK</b> |
| bugs | 3 |
| code_smells | 0 |
| reliability_rating | C (at least 1 Major Bug)<br/>Note: A being best, E being worst |
| security_rating | A (0 Vulnerabilities)<br/>Note: A being best, E being worst |
| vulnerabilities | 0 |

**Comment by asish918** [comment]
| Metric | Value |
| ------ | ----- |
| alert_status | <b>:x: ERROR</b> |
| bugs | 0 |
| code_smells | 13 |
| reliability_rating | A (0 Bugs)<br/>Note: A being best, E being worst |
| security_rating | A (0 Vulnerabilities)<br/>Note: A being best, E being worst |
| vulnerabilities | 0 |

File: backend/homework/5/src/main/java/asish/kdu/q1/Main.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        MessageReceiver receiver2 = new MessageReceiver("Receiver 2", messageQueue);
+        MessageReceiver receiver3 = new MessageReceiver("Receiver 3", messageQueue);
+
+        Thread r1 = new Thread(receiver1);
+        Thread r2 = new Thread(receiver2);
+        Thread r3 = new Thread(receiver3);
+
+        r1.start();
+        r2.start();
+        r3.start();
```
Comment (rishav-kd) [inline backend/homework/5/src/main/java/asish/kdu/q1/Main.java:36]: Lets wrap this inside a try-catch block

File: backend/homework/5/src/main/java/asish/kdu/q1/Main.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+        try {
+            s1.join();
+            s2.join();
+            s3.join();
+            r1.join();
+            r2.join();
+            r3.join();
+        } catch (InterruptedException e) {
+            throw new InterruptedException();
```
Comment (rishav-kd) [inline backend/homework/5/src/main/java/asish/kdu/q1/Main.java:46]: lets not throw an exception here bcoz it is not caught anywhere. lets log it instead.

**Review by rishav-kd — COMMENTED** [review]
Good improvement overall, Used javadocs and mostly all validations are in place.

**Pull request closed without merging**